File: simu_scara_5r/geom_inv.m
function [q1,q2] = geom_inv(x,y)
%% Initialisation des variables
COSPI_4 = sqrt(2)/2;
L1 = 0.065 ;
L2 = 0.08;
D = 0.02;
LT = 0.01;
R_OT_O2 = sqrt( (L2 + LT*sqrt(2)/2)*(L2 + LT*sqrt(2)/2) +  (LT*sqrt(2)/2)*(LT*sqrt(2)/2) );

% % O1 actionneur gauche, O2 actionneur droit, G 2eme pivot gauche D deuxieme
% % pivot droit X organe terminal
% %% Résolution équation intersection deux cercles 
% %Système équation O1-X
% R1 = sqrt((x-O1(1))*(x-O1(1))+(y-O1(2))*(y-O1(2))); % distance entre centres
% ua1 = [(x-O1(1))/R1 (y-O1(2))/R1];%vecteur colinéaire 
% ub1 = [ ua1(2) -ua1(1)]; %vecteur orthogonal
% c1 = [(x+O1(1))/2 (y+O1(2))/2]; % centre segment cercle
% if(R1 > l1+l2)
%     disp('Pas possible O1');
%     return;
% end
% a1 = (l1*l1 - l2*l2)/(2*R1);
% b11 = sqrt((l1*l1+l2*l2)/2 -(l1*l1-l2*l2)*(l1*l1-l2*l2)/(4*R1*R1) -R1*R1/4);
% b12 = -sqrt((l1*l1+l2*l2)/2 -(l1*l1-l2*l2)*(l1*l1-l2*l2)/(4*R1*R1) -R1*R1/4);
% OG1 = c1 + a1*ua1 + b11*ub1; % premiere solution
% OG2 = c1 + a1*ua1 + b12*ub1; % deuxieme solution
% if(OG1(1) < OG2(1)) % prendre solution la plus a gauche
%     OG = OG1;
% else
%     OG = OG2;
% end
% %Système d'équation 02-X
% R2 = sqrt((x-O2(1))*(x-O2(1))+(y-O2(2))*(y-O1(2)));
% ua2 = [(x-O2(1))/R2 (y-O2(2))/R2];
% ub2 = [ ua2(2) -ua2(1)];
% c2 = [(x+O2(1))/2 (y+O2(2))/2];
% if(R2 > l1+l2)
%     disp('Pas possible O2');
%     return;
% end
% a2 = (l1*l1 - l2*l2)/(2*R2);
% b21 = sqrt((l1*l1+l2*l2)/2 -(l1*l1-l2*l2)*(l1*l1-l2*l2)/(4*R2*R2) -R2*R2/4);
% b22 = -sqrt((l1*l1+l2*l2)/2 -(l1*l1-l2*l2)*(l1*l1-l2*l2)/(4*R1*R1) -R2*R2/4);
% OD1 = c2 + a2*ua2 + b21*ub2;
% OD2 = c2 + a2*ua2 + b22*ub2;
% if(OD1(1) < OD2(1)) % prendre solution la plus a droite
%     OD = OG2;
% else
%     OD = OD1;
% end
%% Calcul orientation du modèle
% Création des vecteurs
O1G = [OG(1)-O1(1) OG(2)-O1(2)];
O2D = [OD(1)-O2(1) OD(2)-O2(2)];
GX = [x-OG(1) y-OG(2)];
DX = [x-OD(1) y-OD(2)];
L11 = norm(O1G)
L12 = norm(O2D)
L21 = norm(GX)
L22 = norm(DX)
% Calcul de l'orientation
q1 = acos(dot(O1G,[1 0])/norm(O1G))
q2 = acos(dot(O2D, [1 0])/norm(O2D))

%% Dessin 
th = 0:pi/50:2*pi;
xunit1 = l1 * cos(th) + O1(1);
yunit1 = l1 * sin(th) + O1(2);

xunit2 = l1 * cos(th) + O2(1);
yunit2 = l1 * sin(th) + O2(2);

xunit3 = l2 * cos(th) + x;
yunit3 = l2 * sin(th) + y;

figure;
%Points :
plot(O2(1),O2(2),'o');
hold on;
plot(O1(1),O1(2),'o');
hold on;
plot(OG(1),OG(2),'o');
hold on;
plot(OD(1),OD(2),'o');
hold on;
plot(x,y,'o');
hold on;
% Segments
plot([O1(1) OG(1)],[O1(2) OG(2)],'b');
hold on;
%plot([O1(1) OG1(1)],[O1(2) OG1(2)],'b');
%hold on;

plot([O2(1) OD(1)],[O2(2) OD(2)],'r');
hold on;
%plot([O2(1) OD(1)],[O2(2) OD(2)],'r');
%hold on;
plot([OG(1) x],[OG(2) y],'y');
hold on;
%plot([OG1(1) x],[OG1(2) y],'y');
%hold on;
plot([OD(1) x],[OD(2) y],'black');
hold on;
%plot([OD2(1) x],[OD2(2) y],'black');
%hold on;
% Cercles :
plot(xunit1, yunit1);
hold on;
plot(xunit2, yunit2);
hold on;
plot(xunit3, yunit3);
hold on;
end

File: simu_scara_5r/geom_inv.cpp
#include <iostream>
#include <cmath>

using namespace std;
#define COSPI_4 sqrt(2)/2

void geom_inv(float* X,float l1,float l2,float d,float& q1,float&q2)
{
	// declarations des deux points de controle moteur
	float _O1[2];
	_O1[0] = 0;
	_O1[1] = 0;

	float _O5[2];
	_O5[0] = d;
	_O5[1] = 0;

	float O2[2];
	float O3[2];

	float lt = 0.01;
	float rayon1 = sqrt( pow(l2 + lt*COSPI_4,2) + pow(lt*COSPI_4,2) );
	//Initialisation variables calcul 02
	float _R1 = sqrt( pow((X[0]-_O1[0]),2) + pow((X[1]-_O1[1]),2) ); // Distance 01-X
	// Vecteur colinéaire a R1
	float _ua1[2]; 
	_ua1[0] = (X[0]-_O1[0])/_R1;
	_ua1[1]= (X[1]-_O1[1])/_R1;
	// Vecteur orthogonal a ua1
	float _ub1[2];
	_ub1[0] =  _ua1[1];
	_ub1[1] = -_ua1[0];
	//Point milieu de R1
	float _c1[2];
	_c1[0]= (X[0]+_O1[0])/2;
	_c1[1]= (X[1]+_O1[1])/2;
	//Variable intermédiraires
	float _a1,_b11,_b12;
	float _OG1[2],_OG2[2],_O2[2],_O12[2];

	// Initialisation variable calcul O4
	float _R5 = sqrt( pow((X[0]-_O5[0]),2) + pow((X[1]-_O5[1]),2) ); // Distance O5-X
	//Vecteur collinéaire R5
	float _ua5[2];
	_ua5[0] = (X[0]-_O5[0])/_R5;
	_ua5[1]= (X[1]-_O5[1])/_R5;
	//Vecteur orthogonal
	float _ub5[2];
	_ub5[0] =  _ua5[1];
	_ub5[1] = -_ua5[0];
	//Point milieu 
	float _c5[2];
	_c5[0] = (X[0]+_O5[0])/2;
	_c5[1] = (X[1]+_O5[1])/2;
	//Variables intermédiraires
	float _a5,_b51,_b52;
	float _OD1[2],_OD2[2],_OD[2],_O5D[2];

	//Calcul 02
	if(_R1 > l1+rayon1)
	{
    cout <<"Pas possible O2 \n";
    return;
	}
	else
	{
		_a1 = (l1*l1 - l2*l2)/(2*_R1);
		_b11 = sqrt( (l1*l1 + rayon1*rayon1)/2 -(l1*l1-rayon1*rayon1)*(l1*l1-rayon1*rayon1)/(4*_R1*_R1) -_R1*_R1/4);
		_b12 = -sqrt((l1*l1+rayon1*rayon1)/2 -(l1*l1-rayon1*rayon1)*(l1*l1-rayon1*rayon1)/(4*_R1*_R1) -_R1*_R1/4);
		_OG1[0] = _c1[0] + _a1*_ua1[0] + _b11*_ub1[0]; // premiere solution
		_OG1[1] = _c1[1] + _a1*_ua1[1] + _b11*_ub1[1];
		_OG2[0] = _c1[0] + _a1*_ua1[0] + _b12*_ub1[0]; // deuxieme solution
		_OG2[1] = _c1[1] + _a1*_ua1[1] + _b12*_ub1[1];
		if(_OG1[0] < _OG2[0])
		{ // prendre solution la plus a gauche
    		_O2[0] = _OG1[0];
    		_O2[1] = _OG1[1];
    	}
		else
		{
    		_O2[0] = _OG2[0];
    		_O2[1] = _OG2[1];
    	}
	}
	O12[0] = _O2[0]-O1[0];
	O12[1] = _O2[1]-O1[1];
	q1 = acos(_O12[0]/l1);

	//Initialisation O3
	float _R2 = sqrt( pow((X[0]-_O2[0]),2) + pow((X[1]-_O2[1]),2) ); // Distance 02-X
	// Vecteur colinéaire a R1
	float _ua2[2]; 
	_ua2[0] = (X[0]-_O1[0])/_R3;
	_ua2[1]= (X[1]-_O1[1])/_R3;
	// Vecteur orthogonal a ua1
	float _ub2[2];
	_ub2[0] =  _ua2[1];
	_ub2[1] = -_ua2[0];
	//Point milieu de R1
	float _c2[2];
	_c2[0]= (X[0]+_O2[0])/2;
	_c2[1]= (X[1]+_O2[1])/2;
	//Variable intermédiraires
	float _a2,_b21,_b22;
	float _OG3[2],_OG4[2],_O3[2];


	//Calcul O3
	if(_R2 > l2+lt)
	{
    cout <<"Pas possible O3 \n";
    return;
	}
	else
	{
		_a2 = (l2*l2 - lt*lt)/(2*_R2);
		_b11 = sqrt( (l1*l1 + rayon1*rayon1)/2 -(l1*l1-rayon1*rayon1)*(l1*l1-rayon1*rayon1)/(4*_R1*_R1) -_R1*_R1/4);
		_b12 = -sqrt((l1*l1+rayon1*rayon1)/2 -(l1*l1-rayon1*rayon1)*(l1*l1-rayon1*rayon1)/(4*_R1*_R1) -_R1*_R1/4);
		_OG1[0] = _c1[0] + _a1*_ua1[0] + _b11*_ub1[0]; // premiere solution
		_OG1[1] = _c1[1] + _a1*_ua1[1] + _b11*_ub1[1];
		_OG2[0] = _c1[0] + _a1*_ua1[0] + _b12*_ub1[0]; // deuxieme solution
		_OG2[1] = _c1[1] + _a1*_ua1[1] + _b12*_ub1[1];
		if(_OG1[0] < _OG2[0])
		{ // prendre solution la plus a gauche
    		_O2[0] = _OG1[0];
    		_O2[1] = _OG1[1];
    	}
		else
		{
    		_O2[0] = _OG2[0];
    		_O2[1] = _OG2[1];
    	}
	}
	O12[0] = _O2[0]-O1[0];
	O12[1] = _O2[1]-O1[1];
	q1 = acos(_O12[0]/l1);

	//Initialisation O3
	float _R3 = sqrt( pow((X[0]-_O2[0]),2) + pow((X[1]-_O2[1]),2) ); // Distance 01-X
	// Vecteur colinéaire a R1
	float _ua2[2]; 
	_ua2[0] = (X[0]-_O1[0])/_R3;
	_ua2[1]= (X[1]-_O1[1])/_R3;
	// Vecteur orthogonal a ua1
	float _ub2[2];
	_ub2[0] =  _ua2[1];
	_ub2[1] = -_ua2[0];
	//Point milieu de R1
	float _c2[2];
	_c2[0]= (X[0]+_O2[0])/2;
	_c2[1]= (X[1]+_O2[1])/2;
	//Variable intermédiraires
	float _a2,_b21,_b22;
	float _OG3[2],_OG4[2],_O3[2];




	//Calcul 04
	if(_R5 > l1+l2)
	{
    cout <<"Pas possible O5 \n";
    return;
	}
	else
	{
		_a5 = (l1*l1 - l2*l2)/(2*_R5);
		_b51 = sqrt((l1*l1+l2*l2)/2 -(l1*l1-l2*l2)*(l1*l1-l2*l2)/(4*_R5*_R5) -_R5*_R5/4);
		_b52 = -sqrt((l1*l1+l2*l2)/2 -(l1*l1-l2*l2)*(l1*l1-l2*l2)/(4*_R5*_R5) -_R5*_R5/4);
		_OD1[0] = _c5[0] + _a5*_ua5[0] + _b51*_ub5[0]; // premiere solution
		_OD1[1] = _c5[1] + _a5*_ua5[1] + _b51*_ub5[1];
		_OD2[0] = _c5[0] + _a5*_ua5[0] + _b52*_ub5[0]; // deuxieme solution
		_OD2[1] = _c5[1] + _a5*_ua5[1] + _b12*_ub1[1];
		if(_OD1[0] > _OD2[0])
		{ // prendre solution la plus a droite
    		_OD[0] = _OD1[0];
    		_OD[1] = _OD1[1];
    	}
		else
		{
    		_OD[0] = _OD2[0];
    		_OD[1] = _OD2[1];
    	}
	}
	
	_O5D[0] = _OD[0]-_O5[0];
	_O5D[1] = _OD[1]-_O5[1];
	q1 = acos(_O1G[0]/_norm1);
	q2 = acos(_O5D[0]/_norm2);

}

int main()
{
	float q1 = 0,q2 = 0;
	float X[2];
	X[0] =0.01;
	X[1] =0.1;
	geom_inv(X,0.065,0.080,0.02,q1,q2);
    std::cout << "q1 = " << q1 << ",q2 = " << q2 << '\n';
    return 1;
}